distributed/worker.py

```diff
@@ -920,3 +717,666 @@ def run(worker, stream, function=None, args=(), kwargs={}):
+
+                if key not in self.tasks:
+                    self.data_needed.popleft()
+                    continue
+
+                if self.task_state.get(key) != 'waiting':
+                    self.log.append((key, 'communication pass'))
+                    self.data_needed.popleft()
+                    continue
+
+                deps = self.dependencies[key]
+                deps = [d for d in deps
+                          if d not in self.data
+                          and d not in self.executing
+                          and d not in self.in_flight]
+
+                for dep in deps:
+                    if not self.who_has.get(dep):
+                        logger.info("Can't find dependencies for key %s", key)
+                        self.cancel_key(key)
+                        continue
+
+                self.log.append(('gather-dependencies', key, deps))
+
+                while deps and len(self.connections) < self.total_connections:
+                    self.gather_dep(deps.pop())
+
+                if not deps:
+                    self.data_needed.popleft()
+        except Exception as e:
+            logger.exception(e)
+            if LOG_PDB:
+                import pdb; pdb.set_trace()
+            raise
+
+    def put_key_in_memory(self, key, value):
+        if key in self.data:
+            return
+
+        self.data[key] = value
+
+        if key not in self.nbytes:
+            self.nbytes[key] = sizeof(value)
+
+        for dep in self.dependents.get(key, ()):
+            if dep in self.waiting_for_data:
+                if key in self.waiting_for_data[dep]:
+                    self.waiting_for_data[dep].remove(key)
+                if not self.waiting_for_data[dep]:
+                    self.waiting_for_data[dep]
+                    self.transition(dep, 'ready')
+
+    @gen.coroutine
+    def gather_dep(self, dep):
+        try:
+            if self.validate:
+                self.validate_state()
+
+            while True:
+                if not self.who_has.get(dep):
+                    # TODO: ask scheduler nicely for new who_has before canceling
+                    if dep not in self.dependents:
+                        return
+                    for key in list(self.dependents[dep]):
+                        if dep in self.executing:
+                            continue
+                        if dep in self.waiting_for_data.get(key, ()):
+                            self.cancel_key(key)
+                    return
+                worker = random.choice(list(self.who_has[dep]))
+                ip, port = worker.split(':')
+                try:
+                    future = connect(ip, int(port))
+                    self.connections[future] = True
+                    stream = yield gen.with_timeout(timedelta(seconds=3),
+                                                    future)
+                except AttributeError:  # set_sock_opt sometimes fails
```

Comment:
I'd rather know how to reproduce this. setsockopt() should never fail, or at least it should only fail deterministically.